sdk/include/opentelemetry/sdk/resource/resource.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/sdk/trace/attribute_utils.h"
+#include "opentelemetry/sdk/version/version.h"
+#include "opentelemetry/version.h"
+
+#include <cstdlib>
+#include <memory>
+#include <sstream>
+#include <unordered_map>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace resource
+{
+
+class Resource
+{
+public:
+  Resource(const opentelemetry::common::KeyValueIterable &attributes) noexcept;
+
+  Resource(const opentelemetry::sdk::trace::AttributeMap &attributes) noexcept;
+
+  template <class T,
+            nostd::enable_if_t<common::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Resource(const T &attributes) noexcept : Resource(common::KeyValueIterableView<T>(attributes))
+  {}
+
+  Resource(const std::initializer_list<std::pair<nostd::string_view, common::AttributeValue>>
+               attributes) noexcept
+      : Resource(nostd::span<const std::pair<nostd::string_view, common::AttributeValue>>{
+            attributes.begin(), attributes.end()})
+  {}
+
+  const opentelemetry::sdk::trace::AttributeMap &GetAttributes() const noexcept;
+
+  std::unique_ptr<Resource> Merge(const Resource &other);
+
+  static std::unique_ptr<Resource> Create(
```

Comment:
I see, probably because of creating unique pointers. Should we make the constructors private then, to prevent the creation of unmanaged resources?

Reply:
There is challenge to do this in C++. Ideally, I would like to make constructor as private, and make `OTELResourceDetector` as friend class of `Resource`, but we actually need all derivations of `ResourceDetector` have access to this constructor. 
[opentelementry-java](https://github.com/open-telemetry/opentelemetry-java/blob/f9bf909372a63bd048cfa58a384eba92005b2277/sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java#L30) does it using `AutoValue` annotation, and [opentelemetry-dotnet](https://github.com/open-telemetry/opentelemetry-dotnet/blob/7a91923fb7a1ee6fbc75b6fdc145137e93b38917/src/OpenTelemetry/Resources/Resource.cs#L36) does it using `internal` keyword.

Reply:
Ok. I have now made constructor as protected ( so that it can be used in test cases ), and make `OTELResourceDetector` as friend class. It seems we don't need to invoke `Resource` constructor in any other detector, so we should be fine.